### PR TITLE
discovery(k8s): Only register client-go metrics adapters when needed

### PR DESCRIPTION
**Pull request by heckler1**
Previously the metrics adapters for client-go were registered in an init function. This resulted in clobbering default metrics providers when these packages are imported into an application that leverages the default client-go metrics registry.

Instead, let's only register these adapters when requested.

@brancz for review

### CODE REVIEW

**Comment by bboreham** [comment]
@ptodev you added this code; what do you think?

**Comment by ptodev** [comment]
Hello, the PR looks ok to me. I think the only reason an init() is used right now is to highlight that only the first calls to `RegisterWithK8sGoClient()` will take effect. I'll leave it up to the Prometheus maintainers to decide on the best way to highlight that. 

I personally think it's ok to just leave the PR as it is, since the settings set by `RegisterWithK8sGoClient()` are always the same, and even calling `RegisterK8sClientMetricsWithPrometheus` multiple times with different registries should theoretically work ok.

**Review by roidelapluie — APPROVED** [review]

**Comment by roidelapluie** [comment]
Thanks

**Pull request merged**
